Initial Specification, Now With Forward Secrecy:tm: (#2)
ucan-wg/awake: README.md

# Authorized Wire Authenticated Key Exchange (AWAKE) Specification v0.1.0

## Editors

* [Brooklyn Zelenka](https://github.com/expede), [Fission](https://fission.codes)

## Authors

* [Brooklyn Zelenka](https://github.com/expede), [Fission](https://fission.codes)
* [Daniel Holmgren](https://github.com/dholms), [Bluesky](https://blueskyweb.xyz/)

# 0. Abstract

Authorized Wire Authenticated Key Exchange (AWAKE) is an [AKE](https://en.wikipedia.org/wiki/Authenticated_Key_Exchange) built on top of the [UCAN auth token](https://github.com/ucan-wg/spec). AWAKE is similar to other mutual authentication schemes, such as self-signed [mTLS](https://datatracker.ietf.org/doc/html/rfc8705), but with a focus on authorization proofs. AWAKE leverages the capability chain to prove access to some resource, validating that the requestor is communicating with a party capable of performing certain actions. This is a helpful root of trust with a well defined context when establishing a secure communications channel.

> The core problem that AWAKE solves is bootstrapping a secure session on top of a public channel. Key exchanges for point-to-point communication are plentiful, but in open, trusteless protocols, rooting trust can be a barrier for ad hoc communications channels. Two common approaches are to use a trusted certificate authority, or ignore the principal and "merely" establish a point-to-point channel.

Reviewer: Who is "the principal" here? Is it a certificate authority or some other root of trust?

> Capability-based systems have a helpful philosophy towards a third path. By emphasizing authorization over authentication, they provide a way to know something provable about what the other party "can do", even if they have no sure way of knowing "who they are". One way of phrasing this is that such an agent is "functionally equivalent to the principal in this context".

Reviewer: Yes! The principal makes sense reading this section. It might be good to contextualize the term earlier in the document.

## Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

# 1 Roles

| Name      | Role                                                 |
| --------- | ---------------------------------------------------- |
| Requester | The agent opening the session                        |
| Responder | The agent being contacted by the Requester           |
| Attacker  | An attacker attempting to gain access to the channel |

## 2 Sequence

AWAKE proceeds in one connecion step and four communication rounds:

1. Both parties subscribe to a well-known channel
2. Requester broadcasts intent
    * a. Temporary DID

>     * b. Responder authorization criterea

Reviewer: >    * b. Responder authorization ~criterea~ criteria

Criteria, not criterea.

## Authors

* [Brooklyn Zelenka](https://github.com/expede), [Fission](https://fission.codes)
* [Daniel Holmgren](https://github.com/dholms), [Bluesky](https://blueskyweb.xyz/)

# 0. Abstract

Authorized Wire Authenticated Key Exchange (AWAKE) is an [AKE](https://en.wikipedia.org/wiki/Authenticated_Key_Exchange) built on top of the [UCAN auth token](https://github.com/ucan-wg/spec). AWAKE is similar to other mutual authentication schemes (such as self-signed [mTLS](https://datatracker.ietf.org/doc/html/rfc8705)), but with a focus on authorization and proof. AWAKE leverages the UCAN capability chain to prove access to some resource, validating that the requestor is communicating with a party capable of performing certain actions. This is a helpful root of trust with a well defined context when establishing a secure communications channel.

## Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

# 1 Introduction

> AWAKE bootstraps a secure session on top of a public channel. Key exchanges for point-to-point communication are plentiful, but in open, trusteless protocols, rooting trust can be a barrier for ad hoc communications channels. Two common approaches are to use a trusted certificate authority, or ignore the principal and "merely" establish a point-to-point channel.

Reviewer 2: sp: "trustless"

FIXME add section about KDF in the intro

## 1.1 Payload Fields

All payloads MUST include the "AWAKE Version" field `awv: "0.1.0"`. Payloads MUST also include a message type field `type` (see each stage for the value). All field keys and message type values MUST be case-insensitive.

| Field  | Value          | Description           | Required |
| ------ | -------------- | --------------------- | -------- |
| `awv`  | `"0.1.0"`      | AWAKE message version | Yes      |
| `type` | `awake/<type>` | AWAKE message type    | Yes      |

## 1.2 Roles

| Name      | Role                                                 |
| --------- | ---------------------------------------------------- |
| Requestor | The agent opening the session                        |
| Responder | The agent being contacted by the Requestor           |
| Attacker  | An attacker attempting to gain access to the channel |

## 2 Sequence

AWAKE proceeds in one connection step, four communication rounds, and an OPTIONAL disconnection:

> 1. Both parties subscribe to a well-known channel

Reviewer 2: Is the channel assumed to be long standing?

If not, how often does the requestor broadcast intent?

If the provider joins _after_ the requestor broadcasts intent, when does it get the update that it needs to re-broadcast?

Author: Good questions!

> Is the channel assumed to be long standing?

Yes, it's assumed that it will last at least as long as the handshake

> If not, how often does the requestor broadcast intent?

Whenever they want to start a session (during a session they may exchange a secret to speed up future connections, but not in the protocol)

> If the provider joins after the requestor broadcasts intent, when does it get the update that it needs to re-broadcast?

Yup!

The temporary RSA key from the previous step MUST be exclusively used for exchanging a 256-bit AES "session key". RSA is both slow and can only hold a limited number of bytes, so using it to encrypt the payloads of the rest of the session is infeasable.

The payload contains two encryption layers, and signature: the RSA envelope, the AES envelope, and the AES key signed by the Responder.

```
          Payload

        ┌───RSA───┐
        │         │
        │ AES-GCM │
        │    │    │
        └────┼────┘
             │
             │
             ▼
┌─────────AES-GCM────────┐
│                        │
│  ┌───────UCAN───────┐  │
│  │                  │  │
│  │  iss: Responder  │  │
│  │  aud: TempDID    │  │
│  │  fct: AES-GCM    │  │
│  │  att: []         │  │
│  │  prf: ...        │  │
│  │                  │  │
│  └──────────────────┘  │
│                        │
└────────────────────────┘
```

> 

Reviewer 2: Is there a concern here that some attacker continuously responds to messages before the actual responder?

I'm specifically thinking about regenerating the temp DID here.
- Alice sends out request which is seen by Bob and Eve.
- Eve quickly responds with a bad UCAN
- Alice throws out tempDID & tries again
- repeat until Bob gets a faster internet connection 😛

Author: Oh yeah good point, I should add text that this is nonblocking 👍 (i.e. multiple connections allowed)

Upon receipt, the Requestor MUST validate that the UCAN capabilities fulfill their `caps` criterea. The UCAN itself MUST be valid, unrevoked, unexpired, and intended for the temporary DID (the `aud` field). If any of these checks fail, the session MUST be abandoned, the temporary DID regenerated, and the protocol restarted from [intention braodcast](#32-requestor-broadcasts-intent).

### 3.3.1 Key Exchange

The Responder MUST generate a fresh 256-bit AES key and 12-byte initialization vector (IV) per connection request. It is RECOMMENDED that the Responder track all DIDs requested with, and to only respond to each temporary DID exactly once.

The key used in this step MUST be used as input key material (IKM) to derive keys in rest of the AWAKE bootstrap. The AES key MUST be encoded as padded base64 and included in the facts (`fct`) section of the payload UCAN. The rest of the [validation UCAN is constructed](#332-validation-ucan) and signed per normal, thus including the AES key in the signature payload. This signature serves as proof that the AES key was intended for this specific session.

``` javascript
{
  fct: [
    {

>       "awake/ikm": base64PaddedIkm,

Reviewer 2: Why is this necessary if you already need the key to decrypt it?

Author: There's a reason, but this step is changing substantially due to the double ratchet so 🤷‍♀️

    }
  ]
}
```

The entire UCAN MUST be encrypted with the same AES key as included in the facts section. The GCM operation mode MUST be used.

The encrypted payload MUST have the IV appended prior to encryption. The IV MUST be freshly generated for every message in this session. If the session is not fully established, the AES key MUST be immedietly discarded and MUST NOT ever be reused.

### 3.3.2 Validation UCAN

The validation UCAN MUST NOT be used to delegate any capabilities. This UCAN MUST only be used to prove access to capabilities and sign the AES key. The `att` and `my` fields MUST be empty arrays.

#### 3.3.2.1 Challenge

The Responder MUST set the method of challege to validate the Requestor. This MUST be set in the `fct` section of the UCAN so that it is signed by the Responder. The RECOMMENDED authorization methods are out-of-band PIN validation (`oob-pin`) and UCAN (`ucan`).

To set the challenge as `oob-pin`, the `fct` section of the UCAN MUST include the following:

``` javascript
{
  ...,
  "fct": [
    {"awake/challenge": "oob-pin"}
  ]
}
```

> To set the challenge as `ucan`, the `fct` section of the UCAN MUST include the following:

Reviewer 2: Oh this is interesting. I like that

To set the challenge as `oob-pin`, the `fct` section of the UCAN MUST include the following:

``` javascript
{
  ...,
  "fct": [
    ...,
    {"awake/challenge": "oob-pin"}
  ]
}
```

To set the challenge as `ucan`, the `fct` section of the UCAN MUST include the following:

``` javascript
{
  ...,
  "fct": [
    ...,
    { 
      "awake/challenge": "ucan",
      "caps": [...requiredCaps] 
    }
  ]
}
```

If more than one `awake/challenge` field is set, the lowest-indexed one MUST be used.

#### 3.3.1.2 Next Responder ECDH

The UCAN's facts (`fct`) field MUST also include the next Responder ECDH public key (to be used in Step 4) encoded as `did:key`. Having the next key in the UCAN places it inside the signature envelope, associating the next key with the Responder's UCAN DID.

``` javascript
//JSON encoded
{
  ...,
  "fct": [
    ...,
    {"awake/nextdid": step4EcdhDid}
  ]
}
```

If more than one `awake/nextdid` field is set, the lowest-indexed one MUST be used.

### 3.3.2 Payload

> To start the Double Ratchet, the payload in this stage has the highest number of cleartext fields. Note that the value in the `res` field contain the temporary ECDH DIDs, and MUST NOT use the Responder's actual long-term DID. Conversely, the UCAN inside the encrypted payload MUST use the Responder's long-term DID.

Reviewer: > Note that the value in the `res` field contain the temporary ECDH DIDs

Should `res` be the `iss`? I see `res` in the message type, but not as a field.

Author: It should! Thanks :)

# Authorized Wire Authenticated Key Exchange (AWAKE) Specification v0.1.0

## Editors

* [Brooklyn Zelenka](https://github.com/expede), [Fission](https://fission.codes)

## Authors

* [Daniel Holmgren](https://github.com/dholms), [Bluesky](https://blueskyweb.xyz/)

> * [Quinn Wilton](https://github.com/QuinnWilton), [Fission](https://fission.codes)

Author: cc @QuinnWilton

      │    │        │          │        │            │
      │    └────────┘          └────┬───┘            │
      │                             │                │
      │                             │                │
      │                             ▼                │
      │                        ┌─────────┐  512-603  │
      │                        │         ├───────────┼───► Unique IV
      │                        │  Split  │           │
      │                        │         ├───────────┼───► AES-Key
      │                        └────┬────┘  256-511  │      (OKM)
      │                             │                │
      │                             │ 0-255          │
      │                             │                │
      └─────────────────────────────┼────────────────┘
                                    │
                                    │
                                    ▼
                                  Next
                                 Secret
```

``` javascript
// JS-flavored Pseudocode

const ecdhSecret = ecdh(aliceSk, bobPk)
const awakeTag = 0x4157414B452D5543414E // "AWAKE-UCAN" as hex

> const pseudorandomBits = hkdf.generateBits({ecdh, salt: currentSecret, info: awakeTag, bitLength: 256 + 256 + 12})

Reviewer 3: Should this be `256 + 256 + 96`?

Author: How different are bits and bytes _really?_

Reviewer 3: 7 bits

Author: You're not wrong

# 1 Introduction

AWAKE bootstraps a secure session on top of a public channel. Key exchanges for point-to-point communication are plentiful, but in open, trustless protocols, rooting trust can be a barrier for ad hoc communications channels. Two common approaches are to use a trusted certificate authority, or ignore the principal and "merely" establish a point-to-point channel.

Capability-based systems have a helpful philosophy towards a third path. By emphasizing authorization over authentication, they provide a way to know something provable about what the other party "can do", even if they have no sure way of knowing "who they are". One way of phrasing this is that such an agent is "functionally equivalent to the principal in this context". AWAKE makes use of authorization to bootstrap point-to-point sessions that are both secure and mutually trusted.

## 1.1 Terminology

This document contains shorthand (especially in diagrams) and nuanced senses of some terms. Below is a dictionary of AWAKE-specific terms:

| Term      | Meaning                                              |
| --------- | ---------------------------------------------------- |

> | Attacker  | An attacker attempting to gain access to the channel |

Reviewer 3: Really nitpicky comment, but is it worth avoiding `attacker` in the definition? Just something like:
> A malicious third party attempting to gain access to the channel

| ECDH      | Elliptic Curve Diffie-Hellman                        |
| PK        | Public key                                           |
| Requestor | The agent opening the session                        |
| Responder | The agent being contacted by the Requestor           |
| SK        | Secret (private) key                                 |

## 1.2 Payload Fields

Payloads are encoding agnostic, but JSON is RECOMMENDED. For JSON, any fields that contain non-JSON values (such as ECDH public keys and encryption payloads) MUST be serialized as unpadded [Base64](https://datatracker.ietf.org/doc/html/rfc4648).

> Messages that a peer cannot parse SHOULD be ignored.

Author: Yeah, I can remove that line. The `RECOMMENDED` is there because it's not a _hard_ requirement (e.g. maybe you store-and-forward unreadable messages). Mainly I wanted to convey that it's fine to ignore messages that you can't parse because they're either not meant for you or malicious.

Would you just remove this line?

> Messages that a peer cannot parse SHOULD be ignored.

Reviewer 3: Makes sense! I'd probably either remove it, or change it to `MUST`. Someone can still do whatever they want with the messages, outside of the flow of the protocol, but they shouldn't be processed as part of AWAKE.

### 1.4.1 Asymmetric Keys

UCAN MUST be used as the signature envelope for AWAKE. Any UCAN-compatible asymmetric key MAY be used for signatures, including RSA, Ed25519, P-256, and so on.

The [ECDH portion](#1431-ecdh-input) of the [Double Ratchet KDF](#143-key-derivation) MUST use P-256.

### 1.4.2 Symmetric Keys

All symmetric encryption in AWAKE MUST use [256-bit AES-GCM](https://csrc.nist.gov/publications/detail/sp/800-38d/final). These keys MUST be generated by the [KDF](#143-key-derivation), and SHOULD be non-extractable where possible.

Each encrypted payload MUST include a unique (freshly generated) 12-byte [initialization vector](https://en.wikipedia.org/wiki/Initialization_vector).

### 1.4.3 Key Derivation

AWAKE uses a modified [Extended Triple Diffie-Hellman (X3DH)](https://www.signal.org/docs/specifications/x3dh/) protocol in order to support the WebCrypto API's non-extractable keys and make use of existing UCAN tokens without the need for pre-sharing keys.

> Key derivation in AWAKE's double ratchet MUST use the following algorithm:

Reviewer 3: Is there a reason you're using SHA3 to derive the OKM directly, rather than going through something like [HKDF](https://developer.mozilla.org/en-US/docs/Web/API/HkdfParams)?

I notice that the spec doesn't give any guidance on generating the IV, but with HKDF you'd be able to generate both the OKM and the IV from the output of `ECDH`, and avoid needing to send it along through the protocol. You'd just need to make sure to use the `info` parameter of HKDF to guarantee domain separation for both derivations.

Author: HKDF was used in an earlier draft, because it's available in the WebCrypto API. The short answer is that we can totally use it if it's better. You've gone deeper on this stuff than me, so I'd love to explain the reasoning and get your thoughts!

# What Does Signal Do?

Signal's Double Ratchet paper suggests two systems for the KDF: HMAC (SHA) and HKDF.

# In Favour of HKDF

HKDF intends to be an easy choice to "just use" as a generic KDF. The people that designed it are -- frankly -- much better at cryptography than me!

# Against HKDF

On the flip side, hash chaining HMACs to derive keys is pretty widely used.

HKDF's security comes from SHA2 which is significantly slower than SHA3. HKDF is much more general than a hash function, and does things like expansion:

> 2. To "expand" the generated output of an already reasonably random input such as an existing shared key into a larger cryptographically independent output
> — Wikipedia

In our use case, we need exactly 256-bit outputs, so features like expansion (and the additional computation therein) isn't required. Breaking a SHA3-derived key is the same as reversing the hash, which "should" be intractable (under hash assumptions).

Reviewer 3: > You've gone deeper on this stuff than me, so I'd love to explain the reasoning and get your thoughts!

This also definitely isn't my expertise, and most of what I know is just sourced from cryptographers with far more experience than me!

I don't believe there's anything insecure about using SHA3 here, like you are, but I've got two comments.

> HKDF's security comes from SHA-2 which is significantly slower than SHA-3. 

This sounds backwards to me? I thought SHA-3 was widely considered to be slower, except on specialized hardware. This is a point that's brought up by [Adam Langley](https://www.imperialviolet.org/2017/05/31/skipsha3.html):

> SHA-3 is also [slow](http://bench.cr.yp.to/results-sha3.html), and is even slower than SHA-2 which is already a comparative laggard amongst crypto primitives.

And is also something [David Wong acknowledges](https://cryptologie.net/article/400/maybe-you-shouldnt-skip-sha-3/). He concludes that you should just choose SHA-3 as a general-purpose hashing algorithm anyway, because avoiding the risk of length extension attacks are worth the slowdown, but that's not something we need to worry about with HKDF.

I also [found some hashcat benchmarks](https://gist.github.com/Chick3nman/ccfb883d2d267d94770869b09f5b96ed) someone ran on an M1 earlier this year, and it looks like SHA-2 outperforms SHA-3 there:

```
---------------------------
* Hash-Mode 1400 (SHA2-256)
---------------------------

Speed.#1.........:  1818.9 MH/s (54.07ms) @ Accel:128 Loops:128 Thr:128 Vec:1

----------------------------
* Hash-Mode 17400 (SHA3-256)
----------------------------

Speed.#1.........:   581.7 MH/s (85.14ms) @ Accel:128 Loops:64 Thr:128 Vec:1
```

> In our use case, we need exactly 256-bit outputs, so features like expansion (and the additional computation therein) isn't required.

I'd actually argue that we need 352-bit outputs, because we have a 256-bit key, and a 96-bit IV. Whether or not it's generated as part of the KDF, that IV is needed at some point, so implementers will need to find another 12-bytes each round from somewhere, and I'd argue that baking the generation of the IV into the KDF would both limit possible mistakes elsewhere (using the same IV in both directions, for example).

> I'm weary of pseudorandom IVs, but maybe that's just me?

My understanding of the IV in AES-GCM is that it just needs to be unique, so one of my concerns is that someone will attempt to do so by either using a hardcoded IV (because each party only sends a single message per key), or by using a counter. In both cases they'll be forgetting to take into account that the other party also sends a message, and so would be making use of the same IV.

My thought with using HKDF to derive the IV was to use the sender's DID as the salt to HKDF so that both parties are able to use the same key material to generate a unique IV per round.

> There's a few places in AWAKE that we could be more economical with space, but ease of implementation is the bigger concern for this version.

That makes sense! I agree that saving space isn't the highest priority here, and the larger point behind the suggestion was about removing a degree of freedom from implementers that may lead to IV misuse, with removing that field just being a bonus.

Either way though, this comment definitely isn't me saying that you're doing anything wrong and that it needs to change or anything like that! I was mostly curious about what had led you toward SHA-3 in the first place.

Author: > This sounds backwards to me? I thought SHA-3 was widely considered to be slower, except on specialized hardware

We ran some benchmarks roughly last autumn to test out Wasm, and IIRC the Wasm SHA3-256 outperformed the WebCrypto API's SHA2-256 (not an apples-to-apples comparison, but you'd expect the native function to outperform Wasm). IIRC the SHA3 team also believes this to be the case for pure software implementations.

But that said, we control the stack here, so SHA2 or BLAKE3 or similar are all on the table!

Lemme dig up that post about SHA2 vs SHA3, though.

Author: > My understanding of the IV in AES-GCM is that it just needs to be unique

Okay, yeah, that's true!

> In both cases they'll be forgetting to take into account that the other party also sends a message, and so would be making use of the same IV.

That is a _really good point_ 💯

Reviewer 3: > We ran some benchmarks roughly last autumn to test out Wasm, and IIRC the Wasm SHA3-256 outperformed the WebCrypto API's SHA2-256

Fascinating! I'm seeing the same thing echoed in other benchmarks, like [this](https://github.com/bitauth/libauth/wiki/Benchmarks#browser-sha256-hash-a-32-byte-input) `libauth` benchmark that puts their Wasm sha256 implementation at almost 20x the throughput of `crypto.subtle`, for 32-byte inputs:

```
libauth x 571,012 ops/sec ±2.38% (62 runs sampled)
hash.js x 346,025 ops/sec ±1.31% (63 runs sampled)
crypto.subtle x 37,260 ops/sec ±2.15% (58 runs sampled)
asmcrypto.js x 3,138 ops/sec ±2.38% (20 runs sampled)
```

The PIN values MUST be within the UTF-8 character set. The PIN MUST be included in the `pin` field. It is RECOMMENDED that the PIN be restricted to human-readable characters, and 4 to 10 characters long. If a very long challenge is required, it is RECOMMENDED that the SHA3 hash of the challenge be used rather than putting a large challenge over the wire.

| Field  | Value                                                      | Description                 | Required |
| ------ | ---------------------------------------------------------- | --------------------------- | -------- |
| `did`  |                                                            | "Actual" Requestor DID      | Yes      |
| `sig`  | `sign(requestorPK, sha3_256(responderDid + outOfBandPin))` | Signature of challenge hash | Yes      |

```javascript
{
  "did": requestorActualDid,
  "sig": signedHash
}
```

#### 3.4.2.3 Direct UCAN Challenge

> If UCAN auth is required by the Responder, the Requestor MUST provide a UCAN. This is the same strategy as the one used by the Responder in [§3.3](#33-responder-es tablishes-point-to-point-session): the UCAN MUST be encrypted with the session key and the IV from the enclosing payload, MUST be given in a raw format, and MUST be inline (without a JSON object wrapper or similar).

Reviewer 3: The link to `§3.3` is broken here

Author: Oh I see. The link checker probably returns a `200`, even if that specific anchor is broken. Welp, I guess back to manually checking. Thanks!

Author: We wouldn't have this problem with LaTeX 🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔

Reviewer 3: But only because we'd be too busy fumbling around with spacing.

![image](https://user-images.githubusercontent.com/285821/183711474-25ca6024-39e6-42aa-9452-d4a11108984c.png)

```
Requestor                  Responder
    ⋮                          ⋮
    │           ACK            │ (5)
    │◄─────────────────────────┤
    │                          │
```

### 3.5.1 Payload

| Field  | Value                             | Description                                                    | Required |
| ------ | --------------------------------- | -------------------------------------------------------------- | -------- |
| `awv`  | `"0.1.0"`                         | AWAKE message version                                          | Yes      |
| `type` | `"awake/msg"`                     | Generic AWAKE message type                                     | Yes      |
| `id`   | `sha3_256(resEcdhPk + reqEcdhPk)` | Message ID                                                     | Yes      |
| `iv`   |                                   | Initialization vector for the encrypted payload                | Yes      |
| `msg`  |                                   | Fulfilled challenge payload encrypted with Step 4 ECDH AES-key | Yes      |

#### 3.5.1.1 Encrypted Message

> The encrypted message payload MUST include an `awake/ack` field, with a value of the requestor's long-term DID. This payload MAY contain additional fields. This is often useful if dovetailing the ACK with the first message of a session using the 

Reviewer 3: Did this last sentence get cut off?

Author: ![](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fwww.memesmonkey.com%2Fimages%2Fmemesmonkey%2F59%2F5976e7cfca942a1a9f8b747600dea081.jpeg&f=1&nofb=1)

``` javascript
{
  ...,
  "awake/ack": reqActualDid
}
```

## 4 Secure Session

```
Requestor                  Responder
    ⋮                          ⋮
    │                          │
    │         Messages         │ (6)
    │◄────────────────────────►│
    │                          │
    ⋮                          ⋮
```

Messages sent over an established AWAKE session MUST contain the following keys:
 

> | Field  | Value                                                 | Description                                                             | Required |

Reviewer 3: Nitpick: this table has both a `Required` column, and you say `OPTIONAL` in the descriptions of any non-required fields.

Author: Fair! ✂️

| ------ | ----------------------------------------------------- | ----------------------------------------------------------------------- | -------- |
| `awv`  | `"0.1.0"`                                             | AWAKE message version                                                   | Yes      |
| `type` | `"awake/msg"`                                         | Generic AWAKE message type                                              | Yes      |

> | `mid`  | `sha3_256(latestSenderEcdhPk + latestReceiverEcdhPk)` | Message ID                                                              | Yes      |

Reviewer 3: Is it worth linking to the relevant sections for "Message ID" and "Session ID" in this and the other tables?

| `sid`  | `sha3_256(firstAesKey)`                               | OPTIONAL session ID                                                     | No       |
| `iv`   |                                                       | Initialization vector for the encrypted payload                         | Yes      |
| `msg`  |                                                       | Fulfilled challenge payload encrypted with latest KDF AES-key           | Yes      |
| `esig` | `encrypt(messageKey, sig(senderUcanSk, msg + mid))`   | OPTIONAL message signature using the sender's UCAN-validated public key | No       |

> The OPTIONAL encrypted signature field (`esig`) MAY be included to prove authenticity of the payload. Without this field, an eavesdropper with access to the input secret MAY be able to spoof a message that updates the next ECDH key to one it controls. The signature MUST be encrypted, since some commonly used key types such as RSA allow for the derivation of the signer's public key.

Reviewer 3: I'm having a little bit of trouble visualizing this attack. Do you have a situation in mind where an eavesdropper will have access to the input secret, but without having compromised the device in a way that allows them to produce this signature themselves?

I ask because the `esig` field being optional raises some alarms for me: this is the sort of thing implementers would skip in a first version, so if it's an important mitigation, maybe it shouldn't be optional. And if it actually isn't required, it might be helpful to include a little bit more guidance on making that decision.

Author: haha I had a feel that this was going to come up 😅 

The scenario is that if an attacker manages to spoof a single message, they can _change_ the next ECDH key. Signing with the UCAN DID key avoids this because it authenticates the message without depending on the AES envelope. Signing can't hurt, but also break-in security is the point of the double ratchet.

That said, when writing this section, it did feel like a bridge too far / like I was wearing a stack of tinfoil hats.

You're 100% right that hedging with a SHOULD is the wrong strategy here.

I'll go ahead and drop this field. Thanks for the rubber ducking 🦆

``` javascript
msgId = sha3_256(latestSenderEcdhPk + latestReceiverEcdhPk)
```

The recipient SHOULD calculate the next possible message IDs, based on known keys. Unless a synchronous protocol is being explicitly used, some number of previous keys SHOULD be considered active to receive out-of-order messages. The recipient SHOULD store messages that it cannot match message IDs for.

To protect against a Byzantine peer flooding its connections with a large number of keys, it is RECOMMENDED that the keys have a TTL, be stored in a fixed-size LIFO queue, or both.

## 4.2 Session ID

As out-of-order messages can lead to a large number of messages, an OPTIONAL session ID based on the SHA3 hash of the first KDF output secret (`sha3_256(newSecret)`) MAY be used during the message phase of AWAKE, but moving to a message channel (such as a unique pubsub topic) is RECOMMENDED as it provides the same function with less noise.

# 5 Disconnection

```
Requestor                  Responder
    ⋮                          ⋮
    │                          │
    │           FIN            │ (7)
    │◄────────────────────────►│
    │                          │
    ▀                          ▀ 
```

> Graceful disconnection from an AWAKE attempt MAY be broadcast at any time during an AWAKE session, including to cancel the AWAKE handshake attempt. This payload MUST be the standard [AWAKE message type](#4-secure-session), SHOULD NOT contain any other keys.

Reviewer 3: nitpick: add "and" before "SHOULD NOT"

> This payload MUST be the standard [AWAKE message type](#4-secure-session), and SHOULD NOT contain any other keys.

## 6.2 Unknown Challenge Error

| Field         | Value               | Description                         | Required |
| ------------- | ------------------- | ----------------------------------- | -------- |
| `awake/error` | `unknown-challenge` | Unknown challenge type              | Yes      |
| `awake/mid`   |                     | Message ID that generated the error | Yes      |

``` javascript
// JSON encoded
{
  "awake/error": "unknown-challenge",
  "awake/mid": offendingMessageId
}
```

# 7 Prior Art

## 7.1 Mutual TLS (mTLS)

[mTLS](https://www.rfc-editor.org/rfc/rfc8705.html) is the best-known mutual authentication protocol. In many ways, AWAKE is mTLS with trusted rooted in UCAN and a self-signed capabilities model.

## 7.2 IKEv2

The [Internet Key Exchange (IKE) Protocol](https://datatracker.ietf.org/doc/html/rfc7296) is typically (but not exclusively) used as part of [IPsec](https://en.wikipedia.org/wiki/IPsec). IKE generally uses certificate authorities (CAs). IKE requires that X.509 be supported.

IKE shares many commonalities with AWAKE, including making available of the same cryptographic algorithms (e.g. P-256).

## 7.3 WireGuard

> [WireGuard](https://www.wireguard.com/) is a VPN protocol that is widely deployed via the Linux kernel, and has since been ported to many other systems. It is UDP-based and aimed at raw performance and security. Being so level, is unconstrained in which cryptographic primitives is uses (i.e. Curve25519).

Reviewer: > Being so level, is unconstrained in which cryptographic primitives is uses (i.e. Curve25519).

Should this sentence open with "Being low-level, it is..."? May be something else, but I'm guessing that might be it because it's in the kernel.

Author: Indeed! Fixed 🙏

This step starts the Double Ratchet. The Responder MUST generate a fresh ECDH P-256 key pair. This MUST be combined with the Requestor's ECDH public key to generate a 256-bit AES key, which MUST be used to encrypt the private payload. The Requestor SHOULD accept multiple concurrent connection attempts on this request DID, at least until the handshake is complete.

The payload contains two encryption layers, and signature: the ECDH components, the AES envelope, and the capability proof signed by the Responder's "true" DID.

The SHA2 hash of the AES key generated in this step MUST be used as the first [input secret in the KDF](#1432-secret-input).

```
          Payload

     ┌──────ECDH─────┐
     │               │
     │  256-bit AES  │
     │       │       │
     └───────┼───────┘
             │
             ▼
┌─────────AES-GCM─────────┐
│                         │
│  ┌────────UCAN───────┐  │
│  │                   │  │
│  │  iss: Responder   │  │
│  │  aud: ReqECDH     │  │
│  │  att: []          │  │
│  │  fct: nextResECDH │  │
│  │  prf: ...         │  │
│  │                   │  │
│  └───────────────────┘  │
│                         │
└─────────────────────────┘
```

> Upon receipt, the Requestor MUST validate that the UCAN capabilities fulfill their `caps` criteria. The UCAN itself MUST be valid, unrevoked, unexpired, and intended for the temporary DID (the `aud` field). If any of these checks fail, the session MUST be abandoned, the temporary DID regenerated, and the protocol restarted from [intention broadcast](#32-requestor-broadcasts-intent).

Author: Good suggestion, thanks!

@bgins I recently learned about the new GitHub `suggestion` feature, which may be a handy tool for you too:

```suggestion
Upon receipt, the Requestor MUST validate that the UCAN capabilities in the proof fulfill their caps criteria.
```

Just use `suggestion` as the tag for a multi-line code snippet, just like how you would tripple-backtick and put `js`

```js
const foo = bar
```

> Upon receipt, the Requestor MUST validate that the UCAN capabilities fulfill their `caps` criteria. The UCAN itself MUST be valid, unrevoked, unexpired, and intended for the temporary DID (the `aud` field). If any of these checks fail, the session MUST be abandoned, the temporary DID regenerated, and the protocol restarted from [intention broadcast](#32-requestor-broadcasts-intent).

Reviewer: Hey, that's great! I'll give that a try on my next suggestion. 😄

### 3.3.1 Validation UCAN

> The validation UCAN MUST NOT be used to delegate any capabilities. This UCAN MUST only be used to prove access to capabilities and sign the AES key. The `att` and `my` fields MUST be empty arrays. The issuer (`iss`) field MUST contain the Responder's long-term DID (rather than the temporary ECDH DID). The audience (`aud`) field MUST contain the Requestor's temporary ECDH DID from [§3.2](#32-requestor-broadcasts-intent).

Author: Good catch! Indeed, this is from the previous KEK design 👍

| Field  | Value                                                    | Description                 | Required |
| ------ | -------------------------------------------------------- | --------------------------- | -------- |
| `did`  |                                                          | "Actual" Requestor DID      | Yes      |
| `sig`  | `sign(requestorPK, sha256(responderDid + outOfBandPin))` | Signature of challenge hash | Yes      |

```javascript
{
  "did": requestorActualDid,
  "sig": signedHash
}
```

#### 3.4.2.3 Direct UCAN Challenge

If UCAN auth is required by the Responder, the Requestor MUST provide a UCAN. This is the same strategy as the one used by the Responder in [§3.3](#33-responder-es tablishes-point-to-point-session): the UCAN MUST be encrypted with the session key and the IV from the enclosing payload, MUST be given in a raw format, and MUST be inline (without a JSON object wrapper or similar).

> The UCAN MUST be issued (`iss`) by the Requestor's DID (not the temporary DID), and its audience (`aud`) MUST be the Responder's DID. The `att` field MUST be set to an empty array (i.e. it MUST NOT delegate any capabilities). The `prf` array MUST fulfill the capabilities set by the Responder.

Author: @bgins How about the newly added text just below on line 545?
